[WIP] Azure Update Domain support with the current failure-domain scheduler logic

djsly opened the pull request:
**What this PR does / why we need it**:
In the Azure world, when using Availablity Sets, a VM gets assigned two domains: fault and update. 
Currently, the pod Anti-affinity logic of the scheduler will use some default failure-domains labels as part of scheduling logic. 
One of the failure-domain is the zone, which is set by the cloud-provider. For azure, initially, the zone was set to match the FailureDomain. While this covers unplanned VM disruption, Microsoft could still be reboot VMs of the same Update Domain, for this reason, we should try to have Kubernetes aware of those update domain. 


**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes #40684

**Special notes for your reviewer**:

This is a first take at supporting Update Domain, I expect that this solution might not be best or aligned with the Kubernetes Vision and I'm o with updating/aligning the PR with the best option. 

We are currently about to deploy a large production cluster in Azure, therefore are requiring this Update Domain to be handled properly. 

**Release Notes**

```release-note
```

Commit:
initial try for Update Domain support with the current failure-domain.beta.kubernetes.io zone scheduler logic

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/43581)
<!-- Reviewable:end -->

djsly commented:
I signed it!

djsly commented:
I signed it!

djsly commented:
cc: @colemickens

djsly commented:
I have created a 10 nodes cluster in Azure and updated the following labels manually
```
NAME                 STATUS    AGE       LABELS
kubenode-default-0   Ready     1d        failure-domain.beta.kubernetes.io/zone=FD1UD1
kubenode-default-1   Ready     1d        failure-domain.beta.kubernetes.io/zone=FD1UD0
kubenode-default-2   Ready     1d        failure-domain.beta.kubernetes.io/zone=FD0UD1
kubenode-default-3   Ready     1d        failure-domain.beta.kubernetes.io/zone=FD0UD1
kubenode-default-4   Ready     1d        failure-domain.beta.kubernetes.io/zone=FD0UD0
kubenode-default-5   Ready     1d        failure-domain.beta.kubernetes.io/zone=FD0UD0
kubenode-default-6   Ready     1d        failure-domain.beta.kubernetes.io/zone=FD0UD0
kubenode-default-7   Ready     1d        failure-domain.beta.kubernetes.io/zone=FD1UD0
kubenode-default-8   Ready     1d        failure-domain.beta.kubernetes.io/zone=FD1UD1
kubenode-default-9   Ready     1d        failure-domain.beta.kubernetes.io/zone=FD1UD1
```

When deploying an application with 4 replicas, the 4 instances always get scheduled on 4 nodes with different FD/UP combination

```
NAME                           READY     STATUS    RESTARTS   AGE       IP            NODE
frontend-88237173-7cn2k        1/1       Running   0          24s       172.17.25.7   kubenode-default-3
frontend-88237173-8f4c5        1/1       Running   0          24s       172.17.68.6   kubenode-default-4
frontend-88237173-dp3mr        1/1       Running   0          24s       172.17.19.4   kubenode-default-8
frontend-88237173-fjqvt        1/1       Running   0          24s       172.17.14.6   kubenode-default-7
```

```
NAME                           READY     STATUS    RESTARTS   AGE       IP            NODE
frontend-88237173-l28v9        1/1       Running   0          1m        172.17.88.4   kubenode-default-5
frontend-88237173-mf7rd        1/1       Running   0          1m        172.17.32.6   kubenode-default-2
frontend-88237173-td135        1/1       Running   0          1m        172.17.19.4   kubenode-default-8
frontend-88237173-xvkz5        1/1       Running   0          1m        172.17.64.6   kubenode-default-1
```

```
NAME                           READY     STATUS              RESTARTS   AGE       IP             NODE
frontend-88237173-0k5px        1/1       Running             0          16s       172.17.14.10   kubenode-default-7
frontend-88237173-9l23p        1/1       Running             0          16s       172.17.19.6    kubenode-default-8
frontend-88237173-kg69s        1/1       Running             0          16s       172.17.32.8    kubenode-default-2
frontend-88237173-s2czt        1/1       Running             0          16s       172.17.88.5    kubenode-default-5
```

```
NAME                           READY     STATUS              RESTARTS   AGE       IP            NODE
frontend-88237173-2scph        1/1       Running             0          8s        172.17.64.8   kubenode-default-1
frontend-88237173-flqng        1/1       Running             0          8s        172.17.19.7   kubenode-default-8
frontend-88237173-k0qd0        1/1       Running             0          8s        172.17.32.9   kubenode-default-2
frontend-88237173-vrlft        1/1       Running             0          8s        172.17.88.7   kubenode-default-5
```

colemickens commented:
cc: @brendandburns @khenidak Does this seem sound?

djsly commented:
The test case used was not optimal, when using more FD and UD in an availability set, we will end up having deployments with a small count of replicas all scheduled in the same UP or FD. With the following availability set configuration (FD=3 UP=4), the following three zones will be identified as different zones.

-  FD1UD0
-  FD1UD1
-  FD1UD2

While the zone are indeed different, they are all sharing the same Fault Domain, therefore a 3 replicas deployment could all fall into the same fault domain.

FD | UD | zone
--- | --- | ----
0 | 0 | FD0UD0
0 | 1 | FD0UD1
0 | 2 | FD0UD2
0 | 3 | FD0UD3
1 | 0 | FD1UD0
1 | 1 | FD1UD1
1 | 2 | FD1UD2
1 | 3 | FD1UD3
2 | 0 | FD2UD0
2 | 1 | FD2UD1
2 | 2 | FD2UD2
2 | 3 | FD2UD3

djsly commented:
I think that the best approach is to keep both the UD and FD separated into two different labels and configured the kube-scheduler with the extra label using the `--failure-domain` argument.

The current default list of failure domains is 
- kubernetes.io/hostname
- failure-domain.beta.kubernetes.io/zone
- failure-domain.beta.kubernetes.io/region

We could augment the list in Azure to add 
- failure-domain.azure.microsoft.com/ud

Waiting for your feedback on the new proposed solution.
@colemickens @brendandburns @khenidak

Thanks!

colemickens commented:
ping @brendandburns @khenidak, I'm not really sure what is right to do here. From the info I have, the solution @djsly is proposing makes sense to me. @khenidak I know you've expressed concerns about this elsewhere.

djsly commented:
On my side, I have updated our kubelet systemd service file with a 
````
ExecStartPost=/etc/kubernetes/azure-label-creator.sh
````

where the script contains a logic to populate the label after kubelet starts.

````
#!/bin/bash
UD=$(curl -s http://169.254.169.254/metadata/v1/InstanceInfo | python -c "import sys, json; print json.load(sys.stdin)['UD']")
re='^[0-9]+$'
if ! [[ $UD =~ $re ]] ; then
   echo "error: ($UD) Not a number" >&2; exit 1
fi
echo "Update domain was extracted: $UD"
kubectl label node `hostname` cloud.azure.beta.kubernetes.io/ud=$UD --overwrite=true
````

So far, the scheduling as been pretty good and we seem to be handling the Update Domain failure zone properly. 

I am now waiting for an agreement on the design to start coding this logic into kubelet and the azure-cloud-provider class. 
What I had in mind was to extend the cloud provider interface with a `CloudProviderLabel` Method which would be used to populate a set of labels provided by the cloud provider implementation.

colemickens commented:
@djsly To make sure, in addition to what you just described, you also added the flag to the scheduler to make it consider that new label? Your idea seems sound to me, I'll ping the others internally tomorrow to get agreement.

djsly commented:
Indeed. The --failure-domain arguments was added to the scheduler.  I added the new label to the list of all default labels.  

> On Apr 4, 2017, at 11:19 PM, Cole Mickens <notifications@github.com> wrote:
> 
> @djsly To make sure, in addition to what you just described, you also added the flag to the scheduler to make it consider that new label? Your idea seems sound to me, I'll ping the others internally tomorrow to get agreement.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

djsly commented:
@colemickens have you had a chance to discuss internally about the proposal? I would like to start coding it for this PR today or tomorrow if you guys think it is worth the investment.

khenidak commented:
Thanks @djsly 

Here are some thoughts:

** on K8S ** : 
K8S currently uses zone label for spreading over multiple failure *zones & hostnames*  (only string comparisons used) check https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go#L102 I am not aware of K8S scheduler using a custom cli argument to use additional labels for spreading nor that i am it uses region for spreading.  A quick peak into scheduler code, it does not reveal that is a special cli argument for any specific scheduler https://github.com/kubernetes/kubernetes/blob/master/plugin/cmd/kube-scheduler/app/options/options.go for every thing registered in the scheduler predicates and priorities refer to https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go 

Keep in mind K8S does scheduling over zones on "prioritization" bases.

*side note: the fd/ud are also updated using the cAdvisor, I will actually work with @colemickens to check if we still need this in the bootstrap script*

** On Azure **: 
Azure consists of racks each rack has a top of a rack router (this is a failure domains), each few physical hosts are grouped together to form update domain, this where host os gets patched, updated and so on. The basic rule is simple, only one UD will be patched at a time.

Now, if we used the approach you are proposing we are creating new super concept with distinct zones that will always be equal to UD count * FD Count (as in if we have 3 FDs and 3 UDs, K8S will see distinct 9 zones). Which in reality *of the underlying platform availability* is not the case. As you have have rightly pointed out 3 RC pods can actually fall within FD1UD0, FD1UD1, FD1UD2 (K8S did its job by spreading but spreading all on the same fault domain). 

* What is the impact on users? * The key principal - not the only key principal - to availability on the cloud is to always plan to lose 1 FD & 1 UD of your availability (imagine a situation where Azure is patching 1 UD and one FD fail). That means always users *should* use replica count = min(max(FD)) for the highest availability (and use count of UD = count of FD, and Azure will spread VMs evenly). Keep in mind that K8S will also reschedule pods to healthier nodes.

** Based on the above, I don't think we should merge this PR. **

djsly commented:
@khenidak Thanks for the feedback. I think that my current PR code changes should be ignore as previously mentioned the concatination approach wasn't working. 

For the scheduler the argument is the following
```
[root@kubemaster0 ~]# kube-scheduler --help 2>&1 | grep domain
      --failure-domains string                   Indicate the "all topologies" set for an empty topologyKey when it's used for PreferredDuringScheduling pod anti-affinity. (default "kubernetes.io/hostname,failure-domain.beta.kubernetes.io/zone,failure-domain.beta.kubernetes.io/region")
```
Therefore augmenting the default list with one extra failure-domain

Now... when I looked into this, 1.6 wasn't released yet. Now with 1.6 out... It seems that folks have decided to now use the `failure-domains` argument anymore. 

https://github.com/kubernetes/kubernetes/blob/54f38688d75cf9848ca76a2c542434663b4ef591/plugin/cmd/kube-scheduler/app/options/options.go#L80-L81
```
	fs.StringVar(&s.FailureDomains, "failure-domains", api.DefaultFailureDomains, "Indicate the \"all topologies\" set for an empty topologyKey when it's used for PreferredDuringScheduling pod anti-affinity.")
	fs.MarkDeprecated("failure-domains", "Doesn't have any effect. Will be removed in future version.")
```
I'm not curious to see what they are using for the `PreferredDuringScheduling pod anti-affinity`. I will need to look more into the scheduler code to see how they are planning to manage the extra variables... 

Regarding you comment here

>The key principal - not the only key principal - to availability on the cloud is to always plan to lose 1 >FD & 1 UD of your availability (imagine a situation where Azure is patching 1 UD and one FD fail). >That means always users should use replica count = min(max(FD)) for the highest availability (and >use count of UD = count of FD, and Azure will spread VMs evenly). Keep in mind that K8S will also >reschedule pods to healthier nodes.

The problem we see with this approach of having `count of UD = count of FD` is that currently the max(FD) is set to 3, therefore if we run 9 instances of single threaded processes we loose 33% of our capacity. Therefore, spreading across 20 UD sounds more appealing in certain scenarios. 

Hoping this all make sense.

djsly commented:
CC: @wojtek-t has he was the one who work in the deprecation of the `--failure-domains`` argument.

djsly commented:
@khenidak I took some time to test our the kubernetes cluster deployment using ACS on Azure and I saw that the default FD and UD settings are used. (FD=3, UD=5)

From your proposal, how does Microsoft recommend having a guaranteed HA when using ACS for applications deployed on kubernetes. (if the FD != UD count). From the below screenshot, we have 11 nodes spread across the 3 FDs and 5UDs. 

<img width="1094" alt="screen shot 2017-04-24 at 13 15 27" src="https://cloud.githubusercontent.com/assets/4981802/25349635/87ce9d78-28f0-11e7-80b7-6a923972e7fa.png">

@wojtek-t : is it a fair assumption to say that the scheduler doesn't *by default* use any of the failure-domains labels unless the application manifest defines softPodAntiAffineyRules with a single topologyKey per rules. Therefore we have a spreadout scheduling across the following three key, we should need to have three rules per manifest?

* kubernetes.io/hostname
* failure-domain.beta.kubernetes.io/zone
* failure-domain.beta.kubernetes.io/region

Thanks

djsly commented:
Closing as this approach doesn't work well.
